src/rum/injection.cpp

```diff
@@ -0,0 +1,257 @@
+#include "injection.h"
+
+extern "C" {
+#include <ngx_core.h>
+}
+
+#include <cassert>
+
+#include "datadog_conf.h"
+#include "ngx_http_datadog_module.h"
+#include "string_util.h"
+
+namespace datadog {
+namespace nginx {
+namespace rum {
+namespace {
+
+ngx_table_elt_t *search_header(ngx_http_request_t *request,
+                               std::string_view key) {
+  ngx_list_part_t *part = &request->headers_in.headers.part;
+  auto *h = static_cast<ngx_table_elt_t *>(part->elts);
+
+  for (std::size_t i = 0;; i++) {
+    if (i >= part->nelts) {
+      if (part->next == nullptr) {
+        break;
+      }
+
+      part = part->next;
+      h = static_cast<ngx_table_elt_t *>(part->elts);
+      i = 0;
+    }
+
+    if (key.size() != h[i].key.len ||
+        ngx_strcasecmp((u_char *)key.data(), h[i].key.data) != 0) {
+      continue;
+    }
+
+    return &h[i];
+  }
+
+  return nullptr;
+}
+
+bool is_html_content(ngx_str_t *content_type) {
+  assert(content_type != nullptr);
+  std::string_view content_type_sv = to_string_view(*content_type);
+  return content_type_sv.find("text/html") != content_type_sv.npos;
```

Comment:
Should we inject on `application/xhtml+xml` as well?

Reply:
`xhtml` is a stricter form of html. The browser parse and interpret it as XML. The content inside the `script` tag might be interpreted as XML, to prevent, the JS code must be wrapped in a `CDATA` section. At the moment the injected script is not compliant with xhtml, meaning, we should not inject on xhtml.